api/service/stagedstreamsync/staged_stream_sync.go

```diff
@@ -583,3 +595,133 @@ func (s *StagedStreamSync) EnableStages(ids ...SyncStageID) {
 		}
 	}
 }
+
+func (ss *StagedStreamSync) purgeLastMileBlocksFromCache() {
+	ss.lastMileMux.Lock()
+	ss.lastMileBlocks = nil
+	ss.lastMileMux.Unlock()
+}
+
+// AddLastMileBlock adds the latest a few block into queue for syncing
+// only keep the latest blocks with size capped by LastMileBlocksSize
+func (ss *StagedStreamSync) AddLastMileBlock(block *types.Block) {
+	ss.lastMileMux.Lock()
+	defer ss.lastMileMux.Unlock()
+	if ss.lastMileBlocks != nil {
+		if len(ss.lastMileBlocks) >= LastMileBlocksSize {
+			ss.lastMileBlocks = ss.lastMileBlocks[1:]
+		}
+		ss.lastMileBlocks = append(ss.lastMileBlocks, block)
+	}
+}
+
+func (ss *StagedStreamSync) getBlockFromLastMileBlocksByParentHash(parentHash common.Hash) *types.Block {
+	ss.lastMileMux.Lock()
+	defer ss.lastMileMux.Unlock()
+	for _, block := range ss.lastMileBlocks {
+		ph := block.ParentHash()
+		if bytes.Equal(ph[:], parentHash[:]) {
+			return block
+		}
+	}
+	return nil
+}
+
+func (ss *StagedStreamSync) addConsensusLastMile(bc core.BlockChain, cs *consensus.Consensus) ([]common.Hash, error) {
+	curNumber := bc.CurrentBlock().NumberU64()
+	var hashes []common.Hash
+
+	err := cs.GetLastMileBlockIter(curNumber+1, func(blockIter *consensus.LastMileBlockIter) error {
+		for {
+			block := blockIter.Next()
+			if block == nil {
+				break
+			}
+			if _, err := bc.InsertChain(types.Blocks{block}, true); err != nil {
+				return errors.Wrap(err, "failed to InsertChain")
+			}
+			hashes = append(hashes, block.Header().Hash())
+		}
+		return nil
+	})
+	return hashes, err
+}
+
+func (ss *StagedStreamSync) RollbackLastMileBlocks(ctx context.Context, hashes []common.Hash) error {
+	if len(hashes) == 0 {
+		return nil
+	}
+	utils.Logger().Info().
+		Interface("block", ss.bc.CurrentBlock()).
+		Msg("[STAGED_STREAM_SYNC] Rolling back last mile blocks")
+	if err := ss.bc.Rollback(hashes); err != nil {
```

Comment:
same as previous rollback comment @GheisMohammadi @MaxMustermann2 @Frozen

Reply:
As we discussed, the Rollback function is being used on DNS sync, stream sync, staged sync, command line and so on. So, if any changes is needed for the Rollback function, we will do it on separate PR.